### PR TITLE
Add e2e test for kubebuilder

### PR DESCRIPTION
**Pull request by lichuqiang**
Fix: https://github.com/kubernetes-sigs/kubebuilder/issues/166

the PR
- add e2e framework
- add test case to cover main workflow

TODO: 
-  (in separate PRs) add more tests to cover edge-cases

/assign @droot @pwittrock

### CODE REVIEW

File: test/internal/e2e/kubebuildertest.go [hunk 1]
```diff
@@ -136,14 +178,21 @@ func (kt *KubebuilderTest) CleanUp() error {
...
+}
+
+func (kt *KubebuilderTest) runCommand(cmd *exec.Cmd) ([]byte, error) {
 	cmd.Dir = kt.Dir
 	cmd.Env = os.Environ()
 	command := strings.Join(cmd.Args, " ")
 	output, err := cmd.Output()
 	if err != nil {
-		return fmt.Errorf("%s failed with error: %s", command, string(output))
+		return output, fmt.Errorf("%s failed with error: %v", command, err)
```
Comment (droot) [inline test/internal/e2e/kubebuildertest.go]: Capturing output as part of the error is super useful during the debugging because the error returned by cmd.Output() is generally of the form "Exit 255: ......" and doesnt' contain much info.

File: test/e2e/config.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	testNamespace := installName + "-system"
+
+	return &config{
+		domain:              "example.com" + testSuffix,
+		group:               testGroup,
+		version:             "v1alpha1",
+		kind:                "Foo" + testSuffix,
+		installName:         installName,
+		namespace:           testNamespace,
+		controllerImageName: "gcr.io/kubeships/controller-manager:" + testSuffix,
```
Comment (droot) [inline test/e2e/config.go]: hmm....we will have to pick `project-id` (kubeships) from the gcloud config ?

File: test/e2e/config.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+		workDir:             filepath.Join(framework.TestContext.ProjectDir, "e2e-"+testSuffix),
+	}
+}
+
+type deploymentTemplateArguments struct {
+	Namespace string
+}
+
+var deploymentTemplate = `
```
Comment (droot) [inline test/e2e/config.go]: What is this template and how it is used in the tests ?

File: test/e2e/e2e_test.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+	"testing"
+
+	"github.com/kubernetes-sigs/kubebuilder/test/e2e/framework"
+	"github.com/spf13/pflag"
+)
+
+func init() {
+	framework.RegisterFlags()
+	pflag.Parse()
+}
```
Comment (droot) [inline test/e2e/e2e_test.go]: We should avoid code in the `init` block if possible.

File: test/internal/e2e/kubebuildertest.go (hunk 1 above)
Comment (lichuqiang) [reply test/internal/e2e/kubebuildertest.go]: sounds reasonable

File: test/e2e/config.go (hunk 2 above)
Comment (lichuqiang) [reply test/e2e/config.go]: ah, sorry, here I just copy it from the user guide without much thinking.
In fact the name won't affect anything, as we don't perform image push in the test (as I think it has nothing to do with the project itself).
Let me rename it to something that won't make people confused.

File: test/e2e/config.go (hunk 3 above)
Comment (lichuqiang) [reply test/e2e/config.go]: here I take deployment as an example, to validate core-type resource controller generation, and resource reconciliation

**Comment by lichuqiang** [comment]
> One thing I would like to reconsider is -- the use of ginko for these tests. I have found that ginko adds an additional complexity and I am yet to see its advantages over the standard go test especially after the addition of go subtests https://blog.golang.org/subtests

I'd say it's more out of habit, as most of the golang projects I've been involved in (including k8s and its subprojects) use ginkgo for e2e tests.

as for the complexity  you mentioned, If you mean it may affect on performance, I agree that additional wrapping can have effect on performance. But considering that our project is of relatively small scale, I think we can hardly feel this influence. and if you mean convenience, I think a lot of folks have been so used to it.

Anyway, I'll try to investigate if it's considered necessary.

File: test/e2e/e2e_test.go (hunk 4 above)
Comment (lichuqiang) [reply test/e2e/e2e_test.go]: I'm afraid we have to make the flag things in the `init` block, or the flags we passed in will not be recognized.

**Comment by pwittrock** [comment]
I like ginkgo

**Comment by pwittrock** [comment]
@lichuqiang  this is awesome.  Will review tonight.

File: test/e2e/e2e.go [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+}
+
+var _ = Describe("main workflow", func() {
+	It("should perform main kubebuilder workflow successfully", func() {
+		testSuffix := framework.RandomSuffix()
+		c := initConfig(testSuffix)
+		kubebuilderTest := e2einternal.NewKubebuilderTest(c.workDir, framework.TestContext.BinariesDir)
+
+		prepare(c.workDir)
+		defer cleanup(kubebuilderTest, c.workDir, c.controllerImageName)
```
Comment (pwittrock) [inline test/e2e/e2e.go]: Can this go in TearDown?  If so, would you put a TODO in to move it?

File: test/e2e/e2e.go [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+			status, err := kubebuilderTest.RunKubectlCommand(framework.GetKubectlArgs(getOptions))
+			Expect(err).NotTo(HaveOccurred())
+			if status != "Running" {
+				e = fmt.Errorf("controller pod in %s status", status)
+				return false, nil
+			}
+
+			return true, nil
+		}
+		varifyErr := wait.PollImmediate(500*time.Millisecond, 1*time.Minute, verifyContollerUp)
```
Comment (pwittrock) [inline test/e2e/e2e.go]: why not use `Eventually` from ginkgo?

File: test/e2e/e2e.go [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+			logOutput, err := kubebuilderTest.RunKubectlCommand(framework.GetKubectlArgs(logOptions))
+			Expect(err).NotTo(HaveOccurred())
+
+			if !strings.Contains(logOutput, fmt.Sprintf("to reconcile %s-example", strings.ToLower(c.kind))) {
+				e = fmt.Errorf("created resource object %s-example not reconciled yet", strings.ToLower(c.kind))
+				return false, nil
+			}
+			return true, nil
+		}
+		varifyErr = wait.PollImmediate(500*time.Millisecond, 1*time.Minute, verifyResourceReconciled)
```
Comment (pwittrock) [inline test/e2e/e2e.go]: Same comment about `Eventually`

**Comment by pwittrock** [comment]
@lichuqiang This is awesome!  We should use this to validate the release on Thursday.

Left a couple minor comments.  Fine with doing those as a follow up and merging as is.

File: test/e2e/e2e.go [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+
+		By("creating core-type resource controller")
+		controllerOptions := []string{"--group", "apps", "--version", "v1beta2", "--kind", "Deployment", "--core-type"}
+		err = kubebuilderTest.CreateController(controllerOptions)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("building image")
+		// The scaffold test cases generated for core types cannot work without manually modification.
+		// See https://github.com/kubernetes-sigs/kubebuilder/pull/193 for more details
+		// Skip the test part in build process as we don't care about it here.
```
Comment (pwittrock) [inline test/e2e/e2e.go]: It might be better to simply delete the `go test` for the scaffolded controller that doesn't pass.

File: test/e2e/e2e.go [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+		inputFile := filepath.Join(kubebuilderTest.Dir, "hack", "install.yaml")
+		installOptions := []string{"apply", "-f", inputFile}
+		_, err = kubebuilderTest.RunKubectlCommand(framework.GetKubectlArgs(installOptions))
+		Expect(err).NotTo(HaveOccurred())
+
+		By("validate the controller-manager pod running as expected")
+		verifyContollerUp := func() (bool, error) {
+			e = nil
+			// Get pod name
+			getOptions := []string{"get", "pods", "-n", c.namespace, "-o", "go-template={{ range .items }}{{ if not .metadata.deletionTimestamp }}{{ .metadata.name }}{{ \"\\n\" }}{{ end }}{{ end }}"}
```
Comment (pwittrock) [inline test/e2e/e2e.go]: This surprised me.  At first I thought the string was a template, but it is actually passed to kubectl directly.  Maybe leave a comment saying - Use kubectl to format the output with a go-template

File: test/e2e/e2e.go [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+			podOutput, err := kubebuilderTest.RunKubectlCommand(framework.GetKubectlArgs(getOptions))
+			Expect(err).NotTo(HaveOccurred())
+			// TODO: validate pod replicas if not default to 1
+			podNames := framework.ParseCmdOutput(podOutput)
+			if len(podNames) != 1 {
+				e = fmt.Errorf("expect 1 controller pods running, but got %d", len(podNames))
+				return false, nil
+			}
+			controllerPodName = podNames[0]
```
Comment (pwittrock) [inline test/e2e/e2e.go]: should we check that the controllerPodName has the right name prefix?

File: test/e2e/e2e.go (hunk 9 above)
Comment (pwittrock) [inline test/e2e/e2e.go]: I think kubebuilder applies a label for the controller pods.  We could also use the label selector to ensure we get the right Pod.

File: test/e2e/e2e.go [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+		Expect(err).NotTo(HaveOccurred())
+
+		By("validate the created resource object gets reconciled in controller")
+		verifyResourceReconciled := func() (bool, error) {
+			// Check container log to validate that the created resource object gets reconciled in controller
+			logOptions := []string{"logs", controllerPodName, "-n", c.namespace}
+			logOutput, err := kubebuilderTest.RunKubectlCommand(framework.GetKubectlArgs(logOptions))
+			Expect(err).NotTo(HaveOccurred())
+
+			if !strings.Contains(logOutput, fmt.Sprintf("to reconcile %s-example", strings.ToLower(c.kind))) {
```
Comment (pwittrock) [inline test/e2e/e2e.go]: I think there is a ginkgo matcher we could use for this - e.g. https://onsi.github.io/gomega/#containsubstringsubstr-string-args-interface

File: test/e2e/e2e.go (hunk 6 above)
Comment (pwittrock) [reply test/e2e/e2e.go]: https://onsi.github.io/gomega/#eventually

File: test/e2e/e2e.go [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+	RunSpecs(t, "Kubebuilder e2e suite")
+}
+
+var _ = Describe("main workflow", func() {
+	It("should perform main kubebuilder workflow successfully", func() {
+		testSuffix := framework.RandomSuffix()
+		c := initConfig(testSuffix)
+		kubebuilderTest := e2einternal.NewKubebuilderTest(c.workDir, framework.TestContext.BinariesDir)
+
+		prepare(c.workDir)
```
Comment (pwittrock) [inline test/e2e/e2e.go]: Could this be BeforeSuite?

https://onsi.github.io/ginkgo/#global-setup-and-teardown-beforesuite-and-aftersuite

File: test/e2e/e2e.go (hunk 5 above)
Comment (pwittrock) [reply test/e2e/e2e.go]: https://onsi.github.io/ginkgo/#global-setup-and-teardown-beforesuite-and-aftersuite

**Comment by lichuqiang** [comment]
> We should use this to validate the release on Thursday.
Left a couple minor comments. Fine with doing those as a follow up and merging as is.

Most of the comments make sense to me. I'm working on something urgent about k8s v1.11 today.
So if we mean to use it to the release on Thursday, I'm fine with merging as is, and I'll have an update in separate PR later :)

File: test/e2e/e2e.go (hunk 5 above)
Comment (lichuqiang) [reply test/e2e/e2e.go]: I considered that before, but I guess we can not have input/output in BeforeSuite/AfterSuite func, it's more applicable to the general logic that is not related to a specific case.

and here we need to initialize a list of runtime config params (including things like random work dir) at the beginning of the case, and pass down the config for later use.

File: test/e2e/e2e.go (hunk 6 above)
Comment (lichuqiang) [reply test/e2e/e2e.go]: done

File: test/e2e/e2e.go (hunk 8 above)
Comment (lichuqiang) [reply test/e2e/e2e.go]: done

File: test/e2e/e2e.go (hunk 10 above)
Comment (lichuqiang) [reply test/e2e/e2e.go]: done

File: test/e2e/e2e.go (hunk 9 above)
Comment (lichuqiang) [reply test/e2e/e2e.go]: done

**Comment by pwittrock** [comment]
This looks good as an initial test to me.  Lets have the ginkgo vs not-ginkgo test as a separate conversation as we use ginkgo elsewhere and we should have a consistent approach.  Having *some* tests ginkgo and others not is probably the worst approach.

**Comment by droot** [comment]
@pwittrock fair point. We can have ginkgo vs not-ginkgo as a separate conversation. PR in its current form LGTM.
